### ISSUE TITLE
Constructor injection not working for Providers

### ISSUE THREAD

**Issue by potatopankakes** [ef-labs/vertx-jersey#65]
Jersey is not able to create a `MessageBodyReader` provider when it uses constructor injection.  Code to illustrate the issue is at:  [https://github.com/ef-labs/vertx-jersey/compare/develop...potatopankakes:develop](https://github.com/ef-labs/vertx-jersey/compare/develop...potatopankakes:develop).  It works when lines 22 & 26 of `TestMessageBodyReader.java` are commented out, but it fails with the following call stack:

`
WARNING: HK2 service reification failed for [com.englishtown.vertx.jersey.examples.guice.resources.TestMessageBodyReader] with an exception:
MultiException stack 1 of 2
java.lang.NoSuchMethodException: Could not find a suitable constructor in com.englishtown.vertx.jersey.examples.guice.resources.TestMessageBodyReader class.
	at org.glassfish.jersey.internal.inject.JerseyClassAnalyzer.getConstructor(JerseyClassAnalyzer.java:192)
	at org.jvnet.hk2.internal.Utilities.getConstructor(Utilities.java:178)
	at org.jvnet.hk2.internal.ClazzCreator.initialize(ClazzCreator.java:129)
	at org.jvnet.hk2.internal.ClazzCreator.initialize(ClazzCreator.java:180)
	at org.jvnet.hk2.internal.SystemDescriptor.internalReify(SystemDescriptor.java:740)
	at org.jvnet.hk2.internal.SystemDescriptor.reify(SystemDescriptor.java:694)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.reifyDescriptor(ServiceLocatorImpl.java:461)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.narrow(ServiceLocatorImpl.java:2288)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.access$1200(ServiceLocatorImpl.java:125)
	at org.jvnet.hk2.internal.ServiceLocatorImpl$9.compute(ServiceLocatorImpl.java:1373)
	at org.jvnet.hk2.internal.ServiceLocatorImpl$9.compute(ServiceLocatorImpl.java:1368)
	at org.glassfish.hk2.utilities.cache.internal.WeakCARCacheImpl.compute(WeakCARCacheImpl.java:116)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetAllServiceHandles(ServiceLocatorImpl.java:1430)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getAllServiceHandles(ServiceLocatorImpl.java:1355)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getAllServiceHandles(ServiceLocatorImpl.java:1344)
	at org.glassfish.jersey.internal.inject.Providers.getServiceHandles(Providers.java:354)
	at org.glassfish.jersey.internal.inject.Providers.getProviders(Providers.java:187)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.<init>(MessageBodyFactory.java:222)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.glassfish.hk2.utilities.reflection.ReflectionHelper.makeMe(ReflectionHelper.java:1375)
	at org.jvnet.hk2.internal.ClazzCreator.createMe(ClazzCreator.java:272)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:366)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:487)
	at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:83)
	at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:71)
	at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:97)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:154)
	at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:199)
	at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:122)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2022)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:774)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:737)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:707)
	at org.glassfish.jersey.server.model.ComponentModelValidator.<init>(ComponentModelValidator.java:97)
	at org.glassfish.jersey.server.ApplicationHandler.initialize(ApplicationHandler.java:549)
	at org.glassfish.jersey.server.ApplicationHandler.access$500(ApplicationHandler.java:184)
	at org.glassfish.jersey.server.ApplicationHandler$3.call(ApplicationHandler.java:350)
	at org.glassfish.jersey.server.ApplicationHandler$3.call(ApplicationHandler.java:347)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.processWithException(Errors.java:255)
	at org.glassfish.jersey.server.ApplicationHandler.<init>(ApplicationHandler.java:347)
	at com.englishtown.vertx.jersey.impl.DefaultVertxContainer.getApplicationHandlerDelegate(DefaultVertxContainer.java:101)
	at com.englishtown.vertx.jersey.impl.DefaultVertxContainer.getApplicationHandler(DefaultVertxContainer.java:126)
	at com.englishtown.vertx.jersey.impl.DefaultVertxContainer.start(DefaultVertxContainer.java:49)
	at com.englishtown.vertx.jersey.impl.DefaultJerseyServer.lambda$start$0(DefaultJerseyServer.java:94)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:271)
	at io.vertx.core.impl.TaskQueue.lambda$new$0(TaskQueue.java:60)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
MultiException stack 2 of 2
java.lang.IllegalArgumentException: Errors were discovered while reifying SystemDescriptor(
	implementation=com.englishtown.vertx.jersey.examples.guice.resources.TestMessageBodyReader
	contracts={javax.ws.rs.ext.MessageBodyReader}
	scope=javax.inject.Singleton
	qualifiers={org.glassfish.jersey.internal.inject.Custom}
	descriptorType=CLASS
	descriptorVisibility=NORMAL
	metadata=
	rank=0
	loader=null
	proxiable=null
	proxyForSameScope=null
	analysisName=null
	id=105
	locatorId=1
	identityHashCode=1233967477
	reified=false)
	at org.jvnet.hk2.internal.SystemDescriptor.reify(SystemDescriptor.java:705)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.reifyDescriptor(ServiceLocatorImpl.java:461)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.narrow(ServiceLocatorImpl.java:2288)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.access$1200(ServiceLocatorImpl.java:125)
	at org.jvnet.hk2.internal.ServiceLocatorImpl$9.compute(ServiceLocatorImpl.java:1373)
	at org.jvnet.hk2.internal.ServiceLocatorImpl$9.compute(ServiceLocatorImpl.java:1368)
	at org.glassfish.hk2.utilities.cache.internal.WeakCARCacheImpl.compute(WeakCARCacheImpl.java:116)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetAllServiceHandles(ServiceLocatorImpl.java:1430)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getAllServiceHandles(ServiceLocatorImpl.java:1355)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getAllServiceHandles(ServiceLocatorImpl.java:1344)
	at org.glassfish.jersey.internal.inject.Providers.getServiceHandles(Providers.java:354)
	at org.glassfish.jersey.internal.inject.Providers.getProviders(Providers.java:187)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.<init>(MessageBodyFactory.java:222)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.glassfish.hk2.utilities.reflection.ReflectionHelper.makeMe(ReflectionHelper.java:1375)
	at org.jvnet.hk2.internal.ClazzCreator.createMe(ClazzCreator.java:272)
	at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:366)
	at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:487)
	at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:83)
	at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:71)
	at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:97)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:154)
	at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:199)
	at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:122)
	at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2022)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:774)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:737)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:707)
	at org.glassfish.jersey.server.model.ComponentModelValidator.<init>(ComponentModelValidator.java:97)
	at org.glassfish.jersey.server.ApplicationHandler.initialize(ApplicationHandler.java:549)
	at org.glassfish.jersey.server.ApplicationHandler.access$500(ApplicationHandler.java:184)
	at org.glassfish.jersey.server.ApplicationHandler$3.call(ApplicationHandler.java:350)
	at org.glassfish.jersey.server.ApplicationHandler$3.call(ApplicationHandler.java:347)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.processWithException(Errors.java:255)
	at org.glassfish.jersey.server.ApplicationHandler.<init>(ApplicationHandler.java:347)
	at com.englishtown.vertx.jersey.impl.DefaultVertxContainer.getApplicationHandlerDelegate(DefaultVertxContainer.java:101)
	at com.englishtown.vertx.jersey.impl.DefaultVertxContainer.getApplicationHandler(DefaultVertxContainer.java:126)
	at com.englishtown.vertx.jersey.impl.DefaultVertxContainer.start(DefaultVertxContainer.java:49)
	at com.englishtown.vertx.jersey.impl.DefaultJerseyServer.lambda$start$0(DefaultJerseyServer.java:94)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:271)
	at io.vertx.core.impl.TaskQueue.lambda$new$0(TaskQueue.java:60)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
`

**Comment by adrianluisgonzalez** [comment]
I don't think this is specifically related to the `vertx-jersey` module, but I think your problem is you are missing the `@Inject` annotation on your constructor in `TestMessageBodyReader`.

**Comment by potatopankakes** [comment]
Yes, I noticed that after I posted.  I added the `@Inject` and still get the same results.  I also modified the `examples/inject-vertxrequestprocessor` project which is just `hk2`, and got the same results.  Code is uploaded to my fork.

If this isn't an issue related to `vertx-jersey`, do you have an idea for a workaround?  My full problem is that I need an instance of `Vertx` in the `TestMessageBodyReader`, but for this example, I tried a simple `String` since i wasn't having luck w/injecting a `Vertx` instance.  I noticed in your code, you `bind` an instance of a `WriteStreamBodyWriter`to `MessageBodyWriter`.  I was not successful w/doing that for readers.

**Comment by adrianluisgonzalez** [comment]
Very strange, I tried your fork and the named string was injected after adding the annotation and uncommenting.

**Comment by adrianluisgonzalez** [comment]
[patch.diff.txt](https://github.com/ef-labs/vertx-jersey/files/1338351/patch.diff.txt)

The vertx instance is definitely injectable.  I have a attached a patch with my changes which inject both and have a test for the POST.

Here is the output for mvn install:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.englishtown.vertx.jersey.examples.integration.GuiceIntegrationTest
Starting test: GuiceIntegrationTest#testExceptionMapper 
TestMessageBodyReader - constructor t="With constructor injection" vertx hashCode=1669404901
Http server listening for http://localhost:8080 
TestMessageBodyReader - constructor t="With constructor injection" vertx hashCode=1669404901
Starting test: GuiceIntegrationTest#testGet 
TestMessageBodyReader - constructor t="With constructor injection" vertx hashCode=70243528
Http server listening for http://localhost:8080 
TestMessageBodyReader - constructor t="With constructor injection" vertx hashCode=70243528
Starting test: GuiceIntegrationTest#testPost 
TestMessageBodyReader - constructor t="With constructor injection" vertx hashCode=1887620300
Http server listening for http://localhost:8080 
TestMessageBodyReader - isReadable
TestMessageBodyReader - readFrom
TestMessageBodyReader - constructor t="With constructor injection" vertx hashCode=1887620300
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.397 sec - in com.englishtown.vertx.jersey.examples.integration.GuiceIntegrationTest

Results :

Tests run: 3, Failures: 0, Errors: 0, Skipped: 0
```

**Comment by potatopankakes** [comment]
I must have not rebuilt after adding `@Inject`, because it works for me now as well... both with guice and the hk2.  Sorry, for the goose-chase, Adrian.  I appreciate your time in help me find my mistake.

**Comment by potatopankakes** [comment]
@adrianluisgonzalez, I retried injecting the string in my original code base for the `TestMessageBodyReader`, and it failed to inject.  The `@Inject` was on the constructor.  It was helpful to know that there was not supposed to be issues w/the injection, so I looked elsewhere for the issue.  It turned out to be the 'wrong' `@Inject` annotation.

Throughout my code, I use `import java.inject.Inject`... however, it looks like intelliJ was trying to be helpful in organizing my imports from this:

```
import com.google.inject.Binder;
import com.google.inject.Guice;
import com.google.inject.Injector;
import com.google.inject.Key;
import com.google.inject.Module;

import java.inject.Inject;
```

to this:

```
import com.google.inject.*;

import java.inject.Inject;
```

The intelliJ reformat settings default to using `.*` when one reaches 5 imports from the same package.  Apparently, this made `com.google.inject.Inject` take precedence over `java.inject.Inject`... and that is why my constructor w/the `@Inject` was failing to resolve.  When I switch the imports back and forth, the injection success/failure switches back and forth.
